interface/ceed-elemrestriction.c

```diff
@@ -630,6 +630,52 @@ int CeedElemRestrictionApply(CeedElemRestriction rstr, CeedTransposeMode t_mode,
   return CEED_ERROR_SUCCESS;
 }
 
+/**
+  @brief Restrict an L-vector to an E-vector or apply its transpose ignoring any
+         provided orientations
+
+  @param[in]  rstr    CeedElemRestriction
+  @param[in]  t_mode  Apply restriction or transpose
+  @param[in]  u       Input vector (of size @a l_size when t_mode=@ref CEED_NOTRANSPOSE)
+  @param[out] ru      Output vector (of shape [@a num_elem * @a elem_size] when t_mode=@ref CEED_NOTRANSPOSE).
+                        Ordering of the e-vector is decided by the backend.
+  @param[in]  request Request or @ref CEED_REQUEST_IMMEDIATE
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedElemRestrictionApplyUnsigned(CeedElemRestriction rstr, CeedTransposeMode t_mode, CeedVector u, CeedVector ru, CeedRequest *request) {
+  CeedInt m, n;
+
+  if (t_mode == CEED_NOTRANSPOSE) {
+    m = rstr->num_blk * rstr->blk_size * rstr->elem_size * rstr->num_comp;
+    n = rstr->l_size;
+  } else {
+    m = rstr->l_size;
+    n = rstr->num_blk * rstr->blk_size * rstr->elem_size * rstr->num_comp;
+  }
+  if (n != u->length) {
+    // LCOV_EXCL_START
+    return CeedError(rstr->ceed, CEED_ERROR_DIMENSION,
+                     "Input vector size %" CeedInt_FMT " not compatible with element restriction (%" CeedInt_FMT ", %" CeedInt_FMT ")", u->length, m,
+                     n);
+    // LCOV_EXCL_STOP
+  }
+  if (m != ru->length) {
+    // LCOV_EXCL_START
+    return CeedError(rstr->ceed, CEED_ERROR_DIMENSION,
+                     "Output vector size %" CeedInt_FMT " not compatible with element restriction (%" CeedInt_FMT ", %" CeedInt_FMT ")", ru->length,
+                     m, n);
+    // LCOV_EXCL_STOP
+  }
+  if (rstr->num_elem > 0) {
+    if (rstr->ApplyUnsigned) CeedCall(rstr->ApplyUnsigned(rstr, t_mode, u, ru, request));
+    else CeedCall(rstr->Apply(rstr, t_mode, u, ru, request));
+  }
+  return CEED_ERROR_SUCCESS;
+}
```

Comment:
I'm not sure we want to fallback to `Apply` here? `ApplyUnsigned` is more specific of a request from the caller than `Apply`, so I think we need to make sure we fulfill that specific request. I think we want this fallback to work the other way around:

In that case, by default every backend would implement `ApplyUnsigned`. If `Apply` is not implemented, then `CeedElemRestrictionApply` would fall back to using `ApplyUnsigned`.

Reply:
This is an interesting point. I did not think about this much but my logic for the current fallback structure is the following used in diagonal assembly: User calls `ApplyUnsigned` in this specific case (and nowhere else in the library currently), but since most methods don't have a signed vs. unsigned distinction this just falls back to `Apply`. I imagined people writing new backends would see `ApplyUnsigned` as a special case which typically doesn't require special handling and thus should fall back to the default `Apply`.

Of course I guess the reverse which you propose is the same thing which may make more sense in the literal sense of the methods. Let me know and I'll change it throughout the code if desired.

Reply:
A simpler option that's in the same spirit as my thoughts above would be to delete the else line, add a check for `ApplyUnsigned` at the start of the  function like `Apply` has, and then add lines like
```
  CeedCallBackend(CeedSetBackendFunction(ceed, "ElemRestriction", r, "ApplyUnsigned", CeedElemRestrictionApply_Cuda));
```
to the GPU backends.

This way the backend developer is making an affirmative statement about what their unsigned implementation is rather than letting the interface infer. That would be simpler than my first suggestion while still remaining mildly paranoid about the potential for future errors.

Reply:
Nice. Done in [b17517e](https://github.com/CEED/libCEED/pull/1196/commits/b17517ee98e70bbde11768f0b2de0644e9c68f9a), let me know if I got that right.